src/experimental/utils.cpp

```diff
@@ -0,0 +1,242 @@
+        {
+            /**
+             * Compute intersection between a cell and a laser beam
+             * Arguments:
+                * laser_start [karto::Vector2<kt_double>]: Laser initial point in x and y
+                * laser_end [karto::Vector2<kt_double>]: Laser final point in x and y
+                * robot_grid_pos [karto::Vector2<int>]: Initial grid position in x and y
+                * final_grid_pos [karto::Vector2<int>]: Final grid position in x and y
+                * limit_x [kt_double]: Current cell position in x (Transformed from int to kt_double)
+                * limit_y [kt_double]: Current cell position in y (Transformed from int to kt_double)
+                * resolution [kt_double]: Cell resolution
+             * Return:
+                * std::vector<kt_double>: Intersection point
+             */
+
+            // Cell limits: min_x, max_x, min_y, max_y
+            std::vector<kt_double> cell_limits {limit_x, limit_x + resolution, limit_y, limit_y + resolution};
+
+            // Initial points for each of the 4 corners
+            std::vector<kt_double> initial_x {limit_x, limit_x, limit_x + resolution, limit_x + resolution};
+            std::vector<kt_double> initial_y {limit_y, limit_y, limit_y + resolution, limit_y + resolution};
+
+            // Final points for each of the 4 corners
+            std::vector<kt_double> final_x {limit_x + resolution, limit_x, limit_x + resolution, limit_x};
+            std::vector<kt_double> final_y {limit_y, limit_y + resolution, limit_y, limit_y + resolution};
+
+            // Set the new cell limits
+            updateCellLimits(initial_x, initial_y, final_x, final_y, limit_x, limit_y, cell_limits, robot_grid_pos, final_grid_pos, resolution);
+
+            std::vector<kt_double> inter_x, inter_y;
+
+            /*
+                initial {limit_x, limit_y}
+                final {limit_x + res, limit_y}
+
+                initial {limit_x, limit_y}
+                final {limit_x, limit_y + res}
+
+                initial {limit_x + res, limit_y + res}
+                final {limit_x + res, limit_y}
+
+                initial {limit_x + res, limit_y + res}
+                final {limit_x, limit_y + res}
+            */
+
+            for (int k = 0; k < 4; ++k)
+            {
+                karto::Vector2<kt_double> start{initial_x[k], initial_y[k]};
+                karto::Vector2<kt_double> end{final_x[k], final_y[k]};
+                karto::Vector2<kt_double> intersection = calculateCellIntersectionPoints(laser_start, laser_end, start, end);
+
+                if(intersection.Length() != 0)
+                {
+                    if ((fabs(intersection.GetX()) >= (fabs(cell_limits[0]) - 0.001)) &&
+                    (fabs(intersection.GetX()) <= (fabs(cell_limits[1]) + 0.001)) &&
+                    (fabs(intersection.GetY()) >= (fabs(cell_limits[2]) - 0.001)) &&
+                    (fabs(intersection.GetY()) <= (fabs(cell_limits[3]) + 0.001)))
```

Comment:
@hidmic:

1. Verification is to make sure that intersection points found lies within the current cell limits, because in many cases we find the intersection, but that point is outside our cell of interest.

2. I am using the 0.001 to add some range to the cell limits, during testing I found that sometimes the system were taking the intersection points and sometimes it was not (I think it has something to do with some float numbers) even when the intersection points seemed inside the cell (See the following examples). Adding the 0.001 just allowed me to successfully catch the intersection points inside a cell. Since using magic numbers is not a nice practice, I will define a constant for this purpose, so it is clear it purpose.

Intersection points must be within the cell limits (Note how some points are exactly the same in the comparisons)
```
Intersection points: 20.3066, 20.7
Cell limits: 20.3, 20.4,20.7,20.8
------------------------------------
Intersection points: 20.3934, 20.8
Cell limits: 20.3, 20.4,20.7,20.8
------------------------------------
Intersection points: 20.3071, 20.8
Cell limits: 20.3, 20.4,20.8,20.9
------------------------------------
```